### ISSUE TITLE
noop init container error with cni-plugin is used

### ISSUE THREAD

**Issue by nagaland88** [linkerd/linkerd2#9671]
### What is the issue?

In 2.12.2, this was added: 

Added a noop init container to injected pods when the CNI plugin is enabled
to prevent certain scenarios where a pod can get stuck without an IP address

This is now causing an error when the CNI is used: 

```Init:container has runAsNonRoot and image will run as root ```

### How can it be reproduced?

 Deploy a workload with the CNI plugin enabled having: 
 
 ```
securityContext:
        runAsNonRoot: true
```

### Logs, error output, etc

```
Init:container has runAsNonRoot and image will run as root
```

### output of `linkerd check -o short`

Linkerd extensions checks
=========================

linkerd-jaeger
--------------
‼ jaeger extension proxies are up-to-date
    some proxies are not running the current version:
	* jaeger-injector-6646575865-ddmpr (stable-2.12.1)
    see https://linkerd.io/2.12/checks/#l5d-jaeger-proxy-cp-version for hints
‼ jaeger extension proxies and cli versions match
    jaeger-injector-6646575865-ddmpr running stable-2.12.1 but cli running stable-2.12.2
    see https://linkerd.io/2.12/checks/#l5d-jaeger-proxy-cli-version for hints

Status check results are √

### Environment

- Kubernetes version : v1.22.12-eks
- Linkerd version: 2.12.2

**Comment by mateiidavid** [comment]
@nagaland88 hey 👋🏻  if I understand correctly, you're trying to edit the manifests yourself to add `runAsNonRoot` for the no-op container?

We don't "own" the pause container. It looks as if the error is saying "you're trying to run something as non root, but the image is set to run as user 0". It's possible the image doesn't specify a user to run it as (so it defaults to the root user). We obviously can't make changes to the docker file for the image to embed the uid/guid. 

The pause container itself doesn't do anything, it just sleeps, see this [thread](https://groups.google.com/g/kubernetes-users/c/jVjv0QK4b_o), it's not a real security risk. If you really want to provide the non-root guarantee for your peace of mind, you could try instead to provide a different user id:

```yaml
securityContext:
  runAsNonRoot: true
  runAsUser: <x | x != 0> 
```

We're going to switch the no-op init container with something else for CNI in the next release.

**Comment by davi5e** [comment]
Same here, the Pod Security Context blocks the init container because Linkerd doesn't specify a user...

In our case, we manually removed the noop container so the deployments could start (not ideal, but works).

**Comment by shivjm** [comment]
This is blocking cert-manager’s solver Pods for us. We’ll also revert to the previous stable version for now.

**Comment by Fantaztig** [comment]
Hi, just some more input. This also broke the linkerd helm chart for us as we require resource requests and limits to be set as well as the images to be pulled from specific registries, which is both not possible atm.

**Comment by hawkw** [comment]
PR #9790 should fix this on `stable-2.12.x`, it'll be in the next stable release.

**Comment by NissesSenap** [comment]
So this was a rather boring issue, I would like to increase the scope and also mention securityContext. 

```
  Warning  FailedCreate  0s (x16 over 2m45s)  replicaset-controller  Error creating: admission webhook "validation.gatekeeper.sh" denied the request: [psp-allow-privilege-escalation-container] Privilege escalation container is not allowed: noop
[psp-capabilities] init container <noop> is not dropping all required capabilities. Container must drop all of ["NET_RAW", "CAP_SYS_ADMIN"] or "ALL"
[psp-readonlyrootfilesystem] only read-only root filesystem container is allowed: noop
```

You mention https://github.com/linkerd/linkerd2/issues/9671#issuecomment-1286972557 that the image used will be changed. While we are at it why not drop all securityContext and set some decent user settings?

**Comment by nagaland88** [comment]
@adleong I guess, this is now included in the latest stable version?

**Comment by kleimkuhler** [comment]
@nagaland88 the fix for this issue (#9790) was included in the latest stable release [stable-2.12.3](https://github.com/linkerd/linkerd2/releases/tag/stable-2.12.3).

The change that Alex is talking about—moving from a noop initContainer to a network validator initContainer—was not though. That will continue to be available only on edge releases.

**Comment by joebowbeer** [comment]
In reference to #9678 @kleimkuhler wrote:
> That will continue to be available only on edge releases.

When can I expect network-validator to replace the (insecure) noop pause container in a stable release?
